drivers/gpio/gpio_b91.c

```diff
@@ -0,0 +1,542 @@
+	uint8_t irq = GET_IRQ_NUM(dev);
+	uint8_t status = gpio_b91_irq_en_get(dev);
+
+	gpio_b91_irq_status_clr(irq);
+	gpio_fire_callbacks(&data->callbacks, dev, status);
+}
+#endif
+
+/* API implementation: pin_interrupt_configure */
+static int gpio_b91_pin_interrupt_configure(const struct device *dev,
+					    gpio_pin_t pin,
+					    enum gpio_int_mode mode,
+					    enum gpio_int_trig trig)
+{
+	int ret_status = 0;
+
+	switch (mode) {
+	case GPIO_INT_MODE_DISABLED:                /* GPIO interrupt disable */
+		gpiob_b91_irq_en_clr(dev, pin);
+		break;
+
+	case GPIO_INT_MODE_LEVEL:
+		if (trig == GPIO_INT_TRIG_HIGH) {       /* GPIO interrupt High level */
+			gpio_b91_irq_set(dev, pin, INTR_HIGH_LEVEL);
+		} else if (trig == GPIO_INT_TRIG_LOW) { /* GPIO interrupt Low level */
+			gpio_b91_irq_set(dev, pin, INTR_LOW_LEVEL);
+		} else {
+			ret_status = -ENOTSUP;
+		}
+		break;
+
+	case GPIO_INT_MODE_EDGE:
+		if (trig == GPIO_INT_TRIG_HIGH) {       /* GPIO interrupt Rising edge */
+			gpio_b91_irq_set(dev, pin, INTR_RISING_EDGE);
+		} else if (trig == GPIO_INT_TRIG_LOW) { /* GPIO interrupt Falling edge */
+			gpio_b91_irq_set(dev, pin, INTR_FALLING_EDGE);
+		} else {
+			ret_status = -ENOTSUP;
+		}
+		break;
+
+	default:
+		ret_status = -ENOTSUP;
+		break;
+	}
+
+	return ret_status;
+}
+
+/* API implementation: manage_callback */
+static int gpio_b91_manage_callback(const struct device *dev,
+				    struct gpio_callback *callback,
+				    bool set)
+{
+	struct gpio_b91_data *data = dev->data;
+
+	return gpio_manage_callback(&data->callbacks, callback, set);
+}
+
+/* GPIO driver APIs structure */
+static const struct gpio_driver_api gpio_b91_driver_api = {
+	.pin_configure = gpio_b91_pin_configure,
+	.port_get_raw = gpio_b91_port_get_raw,
+	.port_set_masked_raw = gpio_b91_port_set_masked_raw,
+	.port_set_bits_raw = gpio_b91_port_set_bits_raw,
+	.port_clear_bits_raw = gpio_b91_port_clear_bits_raw,
+	.port_toggle_bits = gpio_b91_port_toggle_bits,
+	.pin_interrupt_configure = gpio_b91_pin_interrupt_configure,
+	.manage_callback = gpio_b91_manage_callback
+};
+
+
+/* GPIO driver registration */
+#define GPIO_B91_INIT(n)						    \
+	static const struct gpio_b91_config gpio_b91_config_##n = {	    \
+		.common = {						    \
+			.port_pin_mask = GPIO_PORT_PIN_MASK_FROM_DT_INST(n) \
+		},							    \
+		.gpio_base = DT_INST_REG_ADDR(n),			    \
+		.irq_num = DT_INST_IRQN(n),				    \
+		.irq_priority = DT_INST_IRQ(n, priority)		    \
+	};								    \
+	static struct gpio_b91_data gpio_b91_data_##n;			    \
+									    \
+	DEVICE_DT_INST_DEFINE(n, gpio_b91_init,				    \
+			      NULL,			    \
+			      &gpio_b91_data_##n,			    \
+			      &gpio_b91_config_##n,			    \
+			      POST_KERNEL,				    \
+			      CONFIG_KERNEL_INIT_PRIORITY_DEFAULT,	    \
+			      &gpio_b91_driver_api);
+
+DT_INST_FOREACH_STATUS_OKAY(GPIO_B91_INIT)
+
+
+/* Initialize interrupt handler for IRQ */
+static void irq_connect(void)
```

Comment:
This is called from gpio_b91_init(), which means it gets run for every single instance, connecting the same IRQ multiple times. I think it would be better to provide a per-instance initialization function that connects the IRQ for that instance. See gpio_mcux.c for an example.

Reply:
Done. Now each instance has its own connect function.
Please pay attention that ``gpio_b91_irq_connect_0..4()`` is not moved to ``DT_NUM_INST_STATUS_OKAY()`` because it is not possible to use ``#if IS_INST_IRQ_EN(n)`` inside ``DT_NUM_INST_STATUS_OKAY(..)``. We should not connect IRQ if it is not selected in DTS.

Reply:
> We have to use `#if IS_INST_IRQ_EN(n)` because we should not connect IRQ if it is not selected in DTS.

I'm confused about this.

I see:

```c
+/* Check that 'inst' has only 1 interrupt selected in dts */
+#define IS_INST_IRQ_EN(inst)    (DT_INST_IRQ_HAS_IDX(inst, 0) != 0) && \
+	(DT_INST_IRQ_HAS_IDX(inst, 1) == 0)
```

But the DTS files mention 3 interrupts, it looks like:

```
		gpioa: gpio@80140300 {
...
			interrupt-parent = <&plic0>;
			interrupts = <25 1>, <26 1>, <27 1>;
...
		};
		plic0: interrupt-controller@e4000000 {
			compatible = "sifive,plic-1.0.0";
			#interrupt-cells = <2>;
			interrupt-controller;
...
		};

```

What's going on here?

Reply:
It works in the next way:

Each GPIO node in **DTSI** (soc-level) file has list of all supported IRQs. It informs users about port capabilities, which IRQ numbers are supported by the port.

```
gpioa: gpio@80140300 {
...
interrupts = <25 1>, <26 1>, <27 1>;
...
};
```
In **DTS** (board-level) file, if user wants to use GPIO port with interrupt, user should select **only one IRQ**:
```
&gpioc {
	interrupts = <25 1>;
	status = "okay";
};
```
On the driver side: driver checks if `interrupts` has **only one node**. It means that user has selected it in the Board's DTS file and really wants to use/config interrupts.

We have such specific implementation for GPIO driver because of the next **silicon limitation**:
We have 5 GPIO ports (port A .. port E) but only 3 IRQ numbers (25 .. 26) which could be assigned for the port.
It means that maximum 3 out of 5 GPIO ports cold be configured to detect interrupts and the decision should be done by user in board-level **DTS** file. It means that we cannot call IRQ_CONNECT for each instance of GPIO driver, we are doing it only for instances which are selected by users in board-level **DTS** file.

@mbolivar-nordic,
For me the current solution looks quite user-friendly and currently I do not see any other implementation options which are more user-friendly.
In case if you have any other ideas please inform me.

Reply:
> On the driver side: driver checks if interrupts has only one node. It means that user has selected it in the Board's DTS file and really wants to use/config interrupts. [...]
>  In case if you have any other ideas please inform me.

Sorry for the slow response; I was on vacation. Can you please explain why checking that `DT_NUM_IRQS(...) == 1` is not what you want?

RIght now, the check is that index 0 exists but index 1 does not. That seems weird given that there are 3 possible indexes.

Reply:
Yes, you are right. Somehow I haven't seen `DT_NUM_IRQS`.
Fixed.
```
#define IS_INST_IRQ_EN(inst)    (DT_NUM_IRQS(DT_DRV_INST(inst)) == 1)
```